### PR TITLE
Remove unused global data descriptor shapes from arguments

### PR DESCRIPTION
**Pull request by tbennun**
This PR allows symbols that exist only in non-transient arrays to generate code successfully. This avoids the case of unused program arguments and enables compilation of more programs, while retaining symbolic analysis.

### CODE REVIEW

File: dace/data.py [hunk 1]
```diff
@@ -689,20 +701,23 @@ def as_arg(self, with_types=True, for_call=False, name=None):
...
         for s in self.strides:
             if isinstance(s, sp.Expr):
                 result |= set(s.free_symbols)
-        if isinstance(self.total_size, sp.Expr):
-            result |= set(self.total_size.free_symbols)
         for o in self.offset:
             if isinstance(o, sp.Expr):
                 result |= set(o.free_symbols)
-
+        if self.transient or all_symbols:
```
Comment (alexnick83) [inline dace/data.py:712]: Why is the transient check only here? What about strides and offset just above?

**Review by alexnick83 — COMMENTED** [review]
Some questions on the selected implementation.

File: dace/sdfg/sdfg.py [hunk 2]
```diff
@@ -1290,14 +1291,22 @@ def free_symbols(self) -> Set[str]:
...
 
-        # Start with the set of SDFG free symbols
-        free_syms |= set(self.symbols.keys())
-
-        # Exclude data descriptor names and constants
+        # Exclude data descriptor names, constants, and shapes of global data descriptors
+        not_strictly_necessary_global_symbols = set()
         for name, desc in self.arrays.items():
             defined_syms.add(name)
+            if not desc.transient:
```
[the remark this answers is not in the document]
Comment (tbennun) [reply dace/sdfg/sdfg.py:1298]: You’re right and I started to rewrite it as such.

File: dace/data.py (hunk 1 above)
Comment (tbennun) [reply dace/data.py:712]: Strides and offset are always necessary if you have any memlet (a[i*stride] will be generated in the code).

File: dace/memlet.py [hunk 3]
```diff
@@ -176,15 +177,16 @@ def to_json(self):
     @staticmethod
     def from_json(json_obj, context=None):
         ret = Memlet()
-        dace.serialize.set_properties_from_json(ret,
-                                                json_obj,
-                                                context=context,
-                                                ignore_properties={'src_subset', 'dst_subset', 'num_accesses', 'is_data_src'})
-        
+        dace.serialize.set_properties_from_json(
```
Comment (alexnick83) [inline dace/memlet.py:180]: Unrelated to the PR, but why is YAPF constantly switching the formatting of such lines?

**Review by alexnick83 — APPROVED** [review]
It appears the ssa-symbol tests didn't need to be amended after all. This makes more sense since this PR should not change the result of `free_symbols`.

**Pull request merged**